Feature/issue 32 matrix exponential

charlesm93 opened the pull request:
#### Submisison Checklist
- [ ] Run unit tests: `./runTests.py test/unit`
  -- ran the unit tests for the new feature I created (both high-level and internal functions)
- [x] Run cpplint: `make cpplint`
- [x] Declare copyright holder and open-source license: see below
#### Summary:

Add matrix_exp() functions: computes the matrix exponential of the input matrix. Uses method from Eigen/unsupported (Pade approximation + scaling and squaring), and a specialized algorithms for 2x2 matrices. 
#### How to Verify:

Uses unit tests:
test/unit/math/prim/mat/fun/matrix_exp_test.cpp
test/unit/math/rev/mat/fun/matrix_exp_test.cpp
test/unit/math/fwd/mat/fun/matrix_exp_test.cpp

Internal functions that support matrix_exp() also have their own unit tests. 
#### Side Effects:

None. 
#### Documentation:

Doxygen.
#### Reviewer Suggestions:

Anybody familiar with matrix exponentials. 
#### Copyright and Licensing

Metrum Research Group LLC

By submitting this pull request, the copyright holder is agreeing to license the submitted work under the following licenses:
- Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
- Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

Note: I used and edited a file from Eigen/unsupported 3.3.0-beta2. The copyright information is written at the top of the file and is unchanged. Said file is stan/math/prim/mat/fun/MatrixExponential.h

syclik commented:
@charlesm93, we're going to have trouble merging this pull request until we get Eigen 3.3 into the code base. If you can't tell, the diff is too big for GitHub to handle.

If you want eyes on this quicker, you can stage this as 2 separate pull requests: the first for updating the library to Eigen 3.3. That should have nothing but a change to the library in it. No other modifications. (We can't actually review that pull request per file.) The second pull request can branch from there and we can more easily see the diffs.

Something that Jenkins caught was a problem with this file:
  stan/math/prim/scal/fun/ldexp.hpp
It should only include `cmath` and not `stan/math/rev/core.hpp`. 

Some additional things for that file:
- the first argument should be `const T& a`, the second argument is fine
- the indentation levels are wrong for the function

I haven't looked at anything else. I just saw that Jenkins failed and I was hunting down why.

bob-carpenter commented:
Yes, please don't roll Eigen library commits in with this branch.

Do you need Eigen 3.3 itself or just its matrix exponential function?  That is, if you take the matrix exponential function from Eigen 3.3, can you compile it with Eigen 3.2.9?   If so, you can just take the matrix exponential function file and modify it and use it to define a regular Stan function.  If you do that, include the Eigen license in the file for the modified function to make its provenance and licensing clear.

charlesm93 commented:
Hi Daniel, hi Bob,

I'll take a closer look tomorrow at the office, but a quick point:

- No, I do NOT need Eigen 3.3. If I added it, it was by mistake. I
recreated the matrix exponential function in the stan::math namespace, and
it should be fully functional with eigen 3.2.9.

I'll edit the branch  first thing tomorrow.

~ Charles

On Sun, Sep 11, 2016 at 10:49 AM, Bob Carpenter <notifications@github.com>
wrote:

> Yes, please don't roll Eigen library commits in with this branch.
>
> Do you need Eigen 3.3 itself or just its matrix exponential function? That
> is, if you take the matrix exponential function from Eigen 3.3, can you
> compile it with Eigen 3.2.9? If so, you can just take the matrix
> exponential function file and modify it and use it to define a regular Stan
> function. If you do that, include the Eigen license in the file for the
> modified function to make its provenance and licensing clear.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/stan-dev/math/pull/390#issuecomment-246184128>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ANUliiDZAFk-VPMFFz-OPEvGtykvPXUKks5qpBTogaJpZM4J5VLn>
> .
>

charlesm93 commented:
I updated the branch by:
- removing eigen 3.3
- correcting ldexp.hpp
- (re-running the unit tests for the matrix_exp with double, rev, and fwd)

I hope this makes the pull request more manageable!

On Sun, Sep 11, 2016 at 5:08 PM, Charles Margossian <charlesm@metrumrg.com>
wrote:

> Hi Daniel, hi Bob,
>
> I'll take a closer look tomorrow at the office, but a quick point:
>
> - No, I do NOT need Eigen 3.3. If I added it, it was by mistake. I
> recreated the matrix exponential function in the stan::math namespace, and
> it should be fully functional with eigen 3.2.9.
>
> I'll edit the branch  first thing tomorrow.
>
> ~ Charles
>
> On Sun, Sep 11, 2016 at 10:49 AM, Bob Carpenter <notifications@github.com>
> wrote:
>
>> Yes, please don't roll Eigen library commits in with this branch.
>>
>> Do you need Eigen 3.3 itself or just its matrix exponential function?
>> That is, if you take the matrix exponential function from Eigen 3.3, can
>> you compile it with Eigen 3.2.9? If so, you can just take the matrix
>> exponential function file and modify it and use it to define a regular Stan
>> function. If you do that, include the Eigen license in the file for the
>> modified function to make its provenance and licensing clear.
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/stan-dev/math/pull/390#issuecomment-246184128>, or mute
>> the thread
>> <https://github.com/notifications/unsubscribe-auth/ANUliiDZAFk-VPMFFz-OPEvGtykvPXUKks5qpBTogaJpZM4J5VLn>
>> .
>>
>
>

bob-carpenter commented on make/libraries:
```diff
@@ -31,8 +31,4 @@ $(CVODES)/lib/libsundials_nvecserial.a: $(SUNDIALS_NVECSERIAL)
-.PHONY: clean-libraries
-clean-libraries:
-	$(RM) $(sort $(SUNDIALS_CVODES) $(SUNDIALS_NVECSERIAL) $(LIBCVODES))
+LIBCVODES = $(CVODES)/lib/libsundials_nvecserial.a $(CVODES)/lib/libsundials_cvodes.a
```
Why is the library clean being removed as part of this pull request? 

Also, should there be a newline at the end of a makefile?

bob-carpenter commented on stan/math/prim/mat/fun/matrix_exp.hpp:
```diff
@@ -0,0 +1,37 @@
+            else
+               B = matrix_exp_compute(A);
+
+            return B;
```
This should all just be

```
return (A.cols() == 2) ? matrix_exp_compute_2x2(A) : matrix_exp_compute(A);
```

bob-carpenter commented on stan/math/prim/mat/fun/matrix_exp_2x2.hpp:
```diff
@@ -0,0 +1,42 @@
+
+        using Eigen::Matrix;
+        using Eigen::Dynamic;
+
```
You shouldn'tt put using statements at namespace level---you have to use the full qualified `Eigen::Matrix` and `Eigen::Dynamic` everywhere.  You can define using statements within function scope if you want.

bob-carpenter commented on stan/math/prim/mat/fun/matrix_exp_2x2.hpp:
```diff
@@ -0,0 +1,42 @@
+            Matrix<T, Dynamic, Dynamic> B;
+            T a = A(0, 0), b = A(0, 1), c = A(1, 0), d = A(1, 1), delta;
+            delta = sqrt((a - d)*(a - d) + 4*b*c);
+
```
Please put spaces around `*` operators.

bob-carpenter commented on stan/math/prim/mat/fun/matrix_exp_2x2.hpp:
```diff
@@ -0,0 +1,42 @@
+            T a = A(0, 0), b = A(0, 1), c = A(1, 0), d = A(1, 1), delta;
+            delta = sqrt((a - d)*(a - d) + 4*b*c);
+
+            B.resize(2, 2);
```
Declare the matrix right before you use it as

```
Matrix<T, Dynamic, Dynamic> B(2, 2);
```

rather than defining it up above and then resizing it.

bob-carpenter commented on stan/math/prim/mat/fun/matrix_exp_2x2.hpp:
```diff
@@ -0,0 +1,42 @@
+            B(1, 0) = 2*c*exp(.5*(a+d)) * sinh(.5*delta);
+            B(1, 1) = exp(.5*(a+d))
+                       * (delta*cosh(.5*delta) + (d-a)*sinh(.5*delta));
+
```
You want to declare local variables for all the repeated subexpressions, like

```
T half_delta = 0.5 * delta;
T cosh_half_delta = cosh(half_delta);
T sinh_half_delta = sinh(half_delta);
T exp_half_a_plus_d = exp(0.5 * (a + d));
```

and then use those below where you have theme repeated.

bob-carpenter commented on stan/math/prim/scal/fun/ldexp.hpp:
```diff
@@ -0,0 +1,25 @@
+      * @param[in] b an integer that is the exponent
+      * @return product of a times 2 to the power b
+      */
+
```
No space between `*/` and start of function.

You want to comment

```
* @tparam T Scalar type of significand.
```

bob-carpenter commented on stan/math/rev/scal/fun/frexp.hpp:
```diff
@@ -0,0 +1,39 @@
+          }
+          void chain() {
+              if (unlikely(boost::math::isnan(avi_->val_)))
+                  avi_->adj_ = std::numeric_limits<double>::quiet_NaN();
```
Just put some doc in here that the derivative is zero for `frexp`.  

We'd really like all of the classes and constructors getting added now to be documented.  We were very lax before and have been trying to catch up in files we touch, too.  It doesn't have to be extensive.

bob-carpenter commented on stan/math/prim/mat/fun/matrix_exp.hpp:
```diff
@@ -0,0 +1,37 @@
+
+            if (A.cols() == 2) B = matrix_exp_compute_2x2(A);
+            else
+               B = matrix_exp_compute(A);
```
Functions shouldn't be called `_compute` --- all functions compute, so it's redundant.  You want to be more descriptive and just use `matrix_exp_2` and `matrix_exp_non2` if the latter also handles the size 1 case, or `matrix_exp_3_plus` if you have another special case for 1 x 1 matrices.  

If you want, you can call them `matrix_exp_2`,  and `matrix_exp_3plus`.  Is the matrix exp of a 1 x 1 also a spec

bob-carpenter commented on stan/math/rev/scal/fun/frexp.hpp:
```diff
@@ -0,0 +1,39 @@
+     * @param a The variable that will be decomposed.
+     * @param b Pointer to an integer that will stored the
+     *          integral power of two.
+     * @return Normalized fraction
```
don't need to indent here.  And you want to indicate in/out.

bob-carpenter commented on test/unit/math/fwd/mat/fun/matrix_exp_compute_test.cpp:
```diff
@@ -0,0 +1,62 @@
+    EXPECT_FLOAT_EQ(-1.471518, output(1,0).d_);
+    EXPECT_FLOAT_EQ(1.103638, output(1,1).d_);
+
+}
```
You want newlines at the end of every code file.

And you want to have a test for a 3 x 3 and bigger inputs here.

bob-carpenter commented on test/unit/math/fwd/mat/fun/matrix_exp_test.cpp:
```diff
@@ -0,0 +1,69 @@
+    
+    EXPECT_THROW(matrix_exp(m1), std::invalid_argument);
+    EXPECT_THROW(matrix_exp(m2), std::invalid_argument);
+}
```
Need some larger than 2 x 2 tests.

bob-carpenter commented on test/unit/math/prim/mat/fun/matrix_exp_compute_test.cpp:
```diff
@@ -0,0 +1,39 @@
+    expect_matrix_eq(m2, matrix_exp_compute(m1));
+}
+
+TEST(MathMatrix, matrix_exp_compute_2) {
```
Why is this called `compute_2` when it has 3 x 3 cases?

bob-carpenter commented on test/unit/math/prim/mat/fun/matrix_exp_compute_test.cpp:
```diff
@@ -0,0 +1,39 @@
+    Matrix<double, Dynamic, Dynamic> m1(3,3), m2(3,3), m3(3,3), I(3,3);
+    m1 << 0, 1, 2, 0, 0, -1, 0, 0, 0;
+    m2 << 1, 1, 1.5, 0, 1, -1, 0, 0, 1;
+    
```
Don't define variables like `m3` or `I` that you don't use.  And push the declarations as far down as you can in the code before you use variables---it makes the code much easier to see becuase it's easier to find declarations.

bob-carpenter commented on test/unit/math/prim/mat/fun/matrix_exp_test.cpp:
```diff
@@ -0,0 +1,46 @@
+    
+    EXPECT_THROW(matrix_exp(m1), std::invalid_argument);
+    EXPECT_THROW(matrix_exp(m2), std::invalid_argument);
+}
```
Tests for 3 x 3 and larger, plus end of line at end of files.

bob-carpenter commented on test/unit/math/rev/scal/fun/frexp_test.cpp:
```diff
@@ -0,0 +1,44 @@
+    template <typename T0>
+    inline T0
+    operator()(const T0& arg1) const {
+        int b;
```
This can't be right as b is never defined, just declared.

bob-carpenter commented on makefile:
```diff
@@ -164,8 +165,7 @@ clean-deps:
 	$(shell find test/prob -name '*_generated_*_test.cpp' -type f -exec rm {} +)
-	$(RM) $(wildcard test/gtest.o test/libgtest* test/prob/generate_tests$(EXE))
-	$(RM) $(wildcard $(CVODES)/lib/*)
+	$(RM) $(wildcard test/gtest.o test/libgtest* test/prob/generate_tests$(EXE))
```
@syclik  Should there be newlines at the end of makefiles?

bob-carpenter commented on stan/math/prim/mat/fun/matrix_exp_2x2.hpp:
```diff
@@ -18,23 +15,27 @@ namespace stan {
+			T delta_cosh = delta * cosh_half_delta;
+			
+			B(0, 0) = exp_half_a_plus_d * (delta_cosh + 
+					  (a - d) * sinh_half_delta);
```
Sorry to be so pedantic on style, but operators should start lines, not end lines and terms should be kept to gether to the extent possible sot hese are easy to read.

Also, you can save `(a - d) * sinh_half_delta` because the `(d - a) * sinh_half_delta` below can just be subtracted.

bob-carpenter commented:
This code looks OK to me, but I'd like someone who knows matrix exponentials to have a look at test coverage.

charlesm93 commented:
Hi Bob,

Still working on your feedback. A few quick comments:
- No edits to make/libraries and makefile are required, I removed all
modifications. Tbh, I'm not even sure why I changed them in the first
place.
- I didn't include a 1x1 specialization, but this could easily be added
later
- In frexp_test, the integer b stores the integral power of two computed by
frexp, and therefore only needs to be declared.
- I will work out 3x3 and higher examples -- I just need examples for which
I can calculate the derivatives exactly.




On Mon, Sep 12, 2016 at 6:36 PM, Bob Carpenter <notifications@github.com>
wrote:

> This code looks OK to me, but I'd like someone who knows matrix
> exponentials to have a look at test coverage.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/stan-dev/math/pull/390#issuecomment-246517487>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ANUlislKNX_kryN8LrgLGlbZWlPxlKxrks5qpdQAgaJpZM4J5VLn>
> .
>

bob-carpenter commented:
Makes sense on the integer return for frexp.

You can also use finite differences for testing.  They're accurate enough for that in most cases. There are built-in finite diff derivative calculators in the math library that can work beyond first-order derivatives.  @rtrangucci has the most experience with those.

charlesm93 commented:
I built a 3x3 test case using the same method I used for the 2x2 case (see
Moler and Van Loan, 2003, section 6).

For the reverse mode autodiff, I got an error testing gradients, when the
expected value was 0. Autodiff computed values such as -1.0913936e-11, and
EXPECT_FLOAT_EQ flagged this as a failure. I see this as a very minor
numerical error.

I switched to EXPECT_NEAR and used a "precision" of 1e-10. This took care
of the 0 gradients. Other gradients, where the value was, say 8.829017,
were however off by more than 1e-10 and failed, even though they had passed
the EXPECT_FLOAT_EQ test. All entries of the matrix passed the test when
the precision was 1e-9. Right now, I'm using EXPECT_NEAR(... ,  ..., 1e-9).

Should matrix_exp fail the test? If not, should the used "precision" for
the unit test be reported or documented somewhere?

On Mon, Sep 12, 2016 at 6:48 PM, Bob Carpenter <notifications@github.com>
wrote:

> Makes sense on the integer return for frexp.
>
> You can also use finite differences for testing. They're accurate enough
> for that in most cases. There are built-in finite diff derivative
> calculators in the math library that can work beyond first-order
> derivatives. @rtrangucci <https://github.com/rtrangucci> has the most
> experience with those.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/stan-dev/math/pull/390#issuecomment-246520112>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ANUliswcaBAw0ytb194FNMR8KAeATFpFks5qpda-gaJpZM4J5VLn>
> .
>

bob-carpenter commented:
On Sep 14, 2016, at 11:54 AM, Charles Margossian <notifications@github.com> wrote:
> 
> I built a 3x3 test case using the same method I used for the 2x2 case (see
> Moler and Van Loan, 2003, section 6).
> 
> For the reverse mode autodiff, I got an error testing gradients, when the
> expected value was 0. Autodiff computed values such as -1.0913936e-11, and
> EXPECT_FLOAT_EQ flagged this as a failure. I see this as a very minor
> numerical error.
> 
> I switched to EXPECT_NEAR and used a "precision" of 1e-10. This took care
> of the 0 gradients.

That's fine.

> Other gradients, where the value was, say 8.829017,
> were however off by more than 1e-10 and failed, even though they had passed
> the EXPECT_FLOAT_EQ test.

You usually want to test relative accuracy, not absolute accuracy.
I don't know if that'd make a difference.

EXPECT_FLOAT_EQ tests relative accuracy, not absolute accuracy.

> All entries of the matrix passed the test when
> the precision was 1e-9. Right now, I'm using EXPECT_NEAR(... , ..., 1e-9).
> 
> Should matrix_exp fail the test? If not, should the used "precision" for
> the unit test be reported or documented somewhere?

You can doc it if you can characterize the accuracy,
but it's not really necessary.  And just testing a few inputs
won't give you a thorough look at precision.

Many of our tests use finite differences and they have very
limited precision.

charlesm93 commented:
>
> EXPECT_FLOAT_EQ tests relative accuracy, not absolute accuracy.


That would make it difficult to test an expected value of 0, especially if
said value is obtained through a numerical approximation.
EXPECT_FLOAT_EQ(0, 1e-25) causes a failure.

I'll rewrite the test to use EXPECT_FLOAT_EQ for non-zero values and
EXPECT_NEAR(..., ..., e-10) for zero values.

This strikes me as a serious defect in google Test though, I might raise an
issue.

On Thu, Sep 15, 2016 at 4:42 PM, Bob Carpenter <notifications@github.com>
wrote:

> On Sep 14, 2016, at 11:54 AM, Charles Margossian <notifications@github.com>
> wrote:
> >
> > I built a 3x3 test case using the same method I used for the 2x2 case
> (see
> > Moler and Van Loan, 2003, section 6).
> >
> > For the reverse mode autodiff, I got an error testing gradients, when the
> > expected value was 0. Autodiff computed values such as -1.0913936e-11,
> and
> > EXPECT_FLOAT_EQ flagged this as a failure. I see this as a very minor
> > numerical error.
> >
> > I switched to EXPECT_NEAR and used a "precision" of 1e-10. This took care
> > of the 0 gradients.
>
> That's fine.
>
> > Other gradients, where the value was, say 8.829017,
> > were however off by more than 1e-10 and failed, even though they had
> passed
> > the EXPECT_FLOAT_EQ test.
>
> You usually want to test relative accuracy, not absolute accuracy.
> I don't know if that'd make a difference.
>
> EXPECT_FLOAT_EQ tests relative accuracy, not absolute accuracy.
>
> > All entries of the matrix passed the test when
> > the precision was 1e-9. Right now, I'm using EXPECT_NEAR(... , ...,
> 1e-9).
> >
> > Should matrix_exp fail the test? If not, should the used "precision" for
> > the unit test be reported or documented somewhere?
>
> You can doc it if you can characterize the accuracy,
> but it's not really necessary. And just testing a few inputs
> won't give you a thorough look at precision.
>
> Many of our tests use finite differences and they have very
> limited precision.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/stan-dev/math/pull/390#issuecomment-247449702>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ANUliur3DtvU2YeFOiIu8JNpMBZLCX6Mks5qqa3MgaJpZM4J5VLn>
> .
>

bgoodri commented on stan/math/prim/mat/fun/matrix_exp_2x2.hpp:
```diff
@@ -0,0 +1,43 @@
+            B(1, 0) = 2 * c * exp_sinh;
+            B(1, 1) = exp_half_a_plus_d * (delta_cosh - ad_sinh_half_delta);
+
+            return 1/delta * B;
```
Can do B / delta here and save an autodiff

bob-carpenter replied:
You can also save 2 * exp_sinh instead of exp_sinh and save another multiply by two.

bob-carpenter replied:
These are less important if you are going to define analytic derivatives, but they'll save a bit of speed even when applied to double values---just that's only done once in a well-written Stan program.

bob-carpenter commented on stan/math/prim/mat/fun/matrix_exp.hpp:
```diff
@@ -0,0 +1,33 @@
+         * @param[in] A A matrix
+         * @return Matrix exponential. 
+         */
+
```
No blank line between doxygen doc and start of function.  Also, you need `@tparam T Scalar type.  And probalby not `A A matrix` but `A Matrix to exponentiate.`

bob-carpenter commented on stan/math/prim/mat/fun/matrix_exp.hpp:
```diff
@@ -0,0 +1,33 @@
+            check_nonzero_size("matrix_exp", "input matrix", A);
+            check_square("matrix_exp", "input matrix", A);
+
+            Eigen::Matrix<T, Eigen::Dynamic, Eigen::Dynamic> B;
```
The variable `B` is declared, butnot used anywhere.  Just eliminate this.

bob-carpenter commented on stan/math/prim/mat/fun/matrix_exp.hpp:
```diff
@@ -0,0 +1,33 @@
+            Eigen::Matrix<T, Eigen::Dynamic, Eigen::Dynamic> B;
+
+            return (A.cols() == 2) ? matrix_exp_2x2(A) : matrix_exp_pade(A);
+       }
```
It's OK to leave it, but you don't need parens around test condition.

bob-carpenter commented on stan/math/prim/mat/fun/matrix_exp_2x2.hpp:
```diff
@@ -0,0 +1,43 @@
+         * Return the matrix exponential of a 2x2 matrix. Reference for 
+         * algorithm: http://mathworld.wolfram.com/MatrixExponential.html
+         *
+         * @param[in] A A 2x2 matrix
```
Doc tparam

bob-carpenter commented on stan/math/prim/mat/fun/matrix_exp_2x2.hpp:
```diff
@@ -0,0 +1,43 @@
+        template <typename T>
+        Eigen::Matrix<T, Eigen::Dynamic, Eigen::Dynamic>
+        matrix_exp_2x2
+          (const Eigen::Matrix<T, Eigen::Dynamic, Eigen::Dynamic>& A) {
```
Put this all on one line---lines should go out to 80 characters max.  If that won't fit, put the open paren `(` right after function and put argument on next line.  But I'm pretty sure this one will fit.

charlesm93 replied:
Didn't fit (85 characters), so moved the `(` up.

bob-carpenter commented on stan/math/rev/scal/fun/frexp.hpp:
```diff
@@ -0,0 +1,43 @@
+  namespace math {
+
+	/*
+	 * The derivative of frexp is 0, because frexp returns
```
This doc should start with a sentence about what the class is for.  Here, it can say something simple, like `Variable implementation for frexp.`  This is what will show up in the one-line summary in the generated doc.

bob-carpenter commented on stan/math/rev/scal/fun/frexp.hpp:
```diff
@@ -0,0 +1,43 @@
+	 */
+      class frexp_vari : public op_v_vari {
+      public:
+          explicit frexp_vari(vari* avi, int* b) :
```
Ideally, the contsructor and chain function will be documented.

bob-carpenter commented on stan/math/rev/scal/fun/frexp.hpp:
```diff
@@ -0,0 +1,43 @@
+     * @return Normalized fraction
+     */
+      inline var frexp(const var& a, int* b) {
+          return var(new frexp_vari(a.vi_, b));
```
We have a two-character indent standard.  So the `return` is too far indented.

charlesm93 commented on makefile:
```diff
@@ -168,4 +168,4 @@ clean-deps:
 	@echo '  removing generated test files'
 	$(shell find test/prob -name '*_generated_*_test.cpp' -type f -exec rm {} +)
-	$(RM) $(wildcard test/gtest.o test/libgtest* test/prob/generate_tests$(EXE))
+	$(RM) $(wildcard test/gtest.o test/libgtest* test/prob/generate_tests$(EXE))
```
Not sure what's happening here. It looks as though the exact same line code was deleted and added. I'm guessing GitHub is comparing revisions, as oppose to final code and current dev version.

syclik replied:
@charlesm93, the difference is removing the newline at the end.

charlesm93 commented on stan/math/prim/mat/fun/MatrixExponential.h:
```diff
@@ -0,0 +1,245 @@
+        template <typename MatrixType>
+        struct matrix_exp_computeUV<MatrixType>
+        {
+        	template <typename T>
```
Yes, this is the code from Eigen/unsupported. I added the last function matrix_exp_pade() and made sure its format followed Stan's standards.

bob-carpenter commented:
I think a newline got removed from the end of the files.

> On Sep 19, 2016, at 10:29 AM, Charles Margossian <notifications@github.com> wrote:
> 
> @charlesm93 commented on this pull request.
> 
> In makefile:
> 
> > @@ -168,4 +168,4 @@ clean-deps:
>  clean-all: clean clean-doxygen clean-deps clean-libraries
>  	@echo '  removing generated test files'
>  	$(shell find test/prob -name '*_generated_*_test.cpp' -type f -exec rm {} +)
> -	$(RM) $(wildcard test/gtest.o test/libgtest* test/prob/generate_tests$(EXE))
> +	$(RM) $(wildcard test/gtest.o test/libgtest* test/prob/generate_tests$(EXE))
> 
> Not sure what's happening here. It looks as though the exact same line code was deleted and added. I'm guessing GitHub is comparing revisions, as oppose to final code and current dev version.
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

bob-carpenter commented:
I don't know how to have lint not check a file.  You can
add 

  // NOLINT

to the end of lines to get it to not check those.

- Bob

> On Sep 19, 2016, at 10:34 AM, Charles Margossian <notifications@github.com> wrote:
> 
> @charlesm93 commented on this pull request.
> 
> In stan/math/prim/mat.hpp:
> 
> > @@ -107,6 +107,8 @@
>  #include <stan/math/prim/mat/fun/log_softmax.hpp>
>  #include <stan/math/prim/mat/fun/log_sum_exp.hpp>
>  #include <stan/math/prim/mat/fun/make_nu.hpp>
> +#include <stan/math/prim/mat/fun/matrix_exp.hpp>
> +#include <stan/math/prim/mat/fun/MatrixExponential.h>
> 
> Ok. Once I do this, cpplint complains about the format of the Eigen/unsupported code. I guess cpplint did not check the .h version, but it's certainly going through the .hpp file. Is there a way of telling cpplint to skip this file? What would be the pros and cons of doing so?
> 
> pro: I don't need to reformat 200 lines of code from Eigen/unsupported
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

bgoodri commented:
I think we need some more tests for the case where the dimension is not small. You might copy some things (not involving complex numbers) from Eigen's tests:
https://bitbucket.org/eigen/eigen/raw/f6382682565c946d46612fe0e36e486bba1371ce/unsupported/test/matrix_exponential.cpp

bob-carpenter commented:
Just make sure you attribute where they came from if you
use the code.  If you just use the numbers, it's no big
deal, but I'd still suggest listing the provenance of the test.

- Bob

> On Sep 23, 2016, at 11:36 AM, bgoodri <notifications@github.com> wrote:
> 
> I think we need some more tests for the case where the dimension is not small. You might copy some things (not involving complex numbers) from Eigen's tests:
> https://bitbucket.org/eigen/eigen/raw/f6382682565c946d46612fe0e36e486bba1371ce/unsupported/test/matrix_exponential.cpp
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

charlesm93 commented:
@bgoodri I looked through their test: I couldn't reproduce the results of the pascal test in C++ nor in R -- I think there may be an error in their code. The inverse test looks promising. I'll add it. 

The test I went for randomly generates a matrix of the form A = S * B * Inverse(S), where B is a diagonal matrix. One can then compute exp(A) = S * exp(diag(B)) * Inverse(S). The code lets you pick the dimension of the matrix, currently 100 x 100.

Matrix_exp passes the test with a few notable exceptions:
- when the expected value is 0, and the pade approximation wanders around e-10, 11, ...
- when the expected value is <e-10, and the relative error becomes high because of numerical errors

I don't think these should be flagged as failures, so I tweaked the unit test to allow a higher relative error for these special cases. We should document these "exceptions" for users. 

The decomposition technique should make it easy to implement tests for fwd and rev (and if not, I'll use finite diff). I wanted to share this approach, in case someone had a major objection (see test/unit/.../mat/.../matrix_exp_test.cpp for implementation).

bob-carpenter commented:
> On Sep 26, 2016, at 2:27 PM, Charles Margossian <notifications@github.com> wrote:
> 
> @bgoodri I looked through their test: I couldn't reproduce the results of the pascal test in C++ nor in R -- I think there may be an error in their code. The inverse test looks promising. I'll add it.
> 
> The test I went for randomly generates a matrix of the form A = S * B * Inverse(S), where B is a diagonal matrix. One can then compute exp(A) = S * exp(diag(B)) * Inverse(S).

Don't ever use inverse.  Use S * exp(diag(B)) / S.  You can
use Stan's mdivide_right function for the division.  Don't know
if it'll be more stable to do the multiplication or division first.

> The code lets you pick the dimension of the matrix, currently 100 x 100.
> 
> Matrix_exp passes the test with a few notable exceptions:
> 
> 	• when the expected value is 0, and the pade approximation wanders around e-10, 11, ...

I think that's just a matter of setting the test up to provide
relative error measurements abs(a - b) / ((abs(a) + abs(b)) / 2) < epsilon
for some epsilon.

> 	• when the expected value is <e-10, and the relative error becomes high because of numerical errors

Any idea why the relative error gets high when the scale is low?
All the relative error should scale with size.  The usual problem
is when you have a bunch of varying scales and try to do things
like subtraction.

> I don't think these should be flagged as failures, so I tweaked the unit test to allow a higher relative error for these special cases. We should document these "exceptions" for users.
> 
> The decomposition technique should make it easy to implement tests for fwd and rev (and if not, I'll use finite diff). I wanted to share this approach, in case someone had a major objection (see test/unit/.../mat/.../matrix_exp_test.cpp for implementation).

Yes, at least finite diffs.

- Bob

charlesm93 commented:
>
> Don't ever use inverse.


What's wrong with inverse?

Any idea why the relative error gets high when the scale is low?


I don't think the scale is low, rather some entries are small relative to
other entries. If I look at exp(A), most entries have an order of magnitude
around 0 or 1, with some exceptions at e-12, e-13, etc.

Yes, at least finite diffs.


Do you mean you have an objection with using finite diffs, or that the
gradient tests should at least use finite diffs?

On Mon, Sep 26, 2016 at 2:34 PM, Bob Carpenter <notifications@github.com>
wrote:

>
> > On Sep 26, 2016, at 2:27 PM, Charles Margossian <
> notifications@github.com> wrote:
> >
> > @bgoodri I looked through their test: I couldn't reproduce the results
> of the pascal test in C++ nor in R -- I think there may be an error in
> their code. The inverse test looks promising. I'll add it.
> >
> > The test I went for randomly generates a matrix of the form A = S * B *
> Inverse(S), where B is a diagonal matrix. One can then compute exp(A) = S *
> exp(diag(B)) * Inverse(S).
>
> Don't ever use inverse. Use S * exp(diag(B)) / S. You can
> use Stan's mdivide_right function for the division. Don't know
> if it'll be more stable to do the multiplication or division first.
>
> > The code lets you pick the dimension of the matrix, currently 100 x 100.
> >
> > Matrix_exp passes the test with a few notable exceptions:
> >
> > • when the expected value is 0, and the pade approximation wanders
> around e-10, 11, ...
>
> I think that's just a matter of setting the test up to provide
> relative error measurements abs(a - b) / ((abs(a) + abs(b)) / 2) < epsilon
> for some epsilon.
>
> > • when the expected value is <e-10, and the relative error becomes high
> because of numerical errors
>
> Any idea why the relative error gets high when the scale is low?
> All the relative error should scale with size. The usual problem
> is when you have a bunch of varying scales and try to do things
> like subtraction.
>
> > I don't think these should be flagged as failures, so I tweaked the unit
> test to allow a higher relative error for these special cases. We should
> document these "exceptions" for users.
> >
> > The decomposition technique should make it easy to implement tests for
> fwd and rev (and if not, I'll use finite diff). I wanted to share this
> approach, in case someone had a major objection (see
> test/unit/.../mat/.../matrix_exp_test.cpp for implementation).
>
> Yes, at least finite diffs.
>
> - Bob
>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/stan-dev/math/pull/390#issuecomment-249657118>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ANUlig6yI_x0iYFDVDpp8zx28ROmjwi1ks5quBAzgaJpZM4J5VLn>
> .
>

charlesm93 commented:
I added unit tests for fwd and rev mode.

The code is non-trivial, but the matrix_exp and its derivatives are compared to an exact value. Should I add descriptive comments or a doc? Everything is based on using matrices of the form A = S * B * Inverse(S). 

For fwd and rev test, I set the size of the input matrix to 10x10, to make the code run faster (particularly for rev). I could scale it up, if we want to test higher dimensions.

bob-carpenter commented:
> On Sep 26, 2016, at 2:56 PM, Charles Margossian <notifications@github.com> wrote:
> 
> >
> > Don't ever use inverse.
> 
> 
> What's wrong with inverse?

It's unstable numerically compared to matrix division.
And usually you are multiplying the inverse of a matrix by
a vector or matrix, so division is a viable alternative.

> Any idea why the relative error gets high when the scale is low?

I don't know how they measure relative error or what the algorithm
is.  I'm also not sure what the right thing to do is.

> I don't think the scale is low, rather some entries are small relative to
> other entries. If I look at exp(A), most entries have an order of magnitude
> around 0 or 1, with some exceptions at e-12, e-13, etc.
> 
> Yes, at least finite diffs.
> 
> 
> Do you mean you have an objection with using finite diffs, or that the
> gradient tests should at least use finite diffs?

No, I mean that you should use finite diffs to test if you
don't have an independent answer.

- Bob

bob-carpenter commented:
How slow is it to test larger matrices?  

I also don't know if that's a general enough test.  You
mean that the matrices you test are generated as S * B / S
for some S and B?

- Bob

> On Sep 27, 2016, at 1:45 PM, Charles Margossian <notifications@github.com> wrote:
> 
> I added unit tests for fwd and rev mode.
> 
> The code is non-trivial, but the matrix_exp and its derivatives are compared to an exact value. Should I add descriptive comments or a doc? Everything is based on using matrices of the form A = S * B * Inverse(S).
> 
> For fwd and rev test, I set the size of the input matrix to 10x10, to make the code run faster (particularly for rev). I could scale it up, if we want to test higher dimensions.
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

charlesm93 commented:
@bob-carpenter : I don't have a solid theory for what happens when the scale goes down, but I'm working on it. For now, here's a simple example where the error occurs, using 2x2 matrices:

A = (-1.97e-13, 9.9e-14, 5.94e-13, 2.98e-13)

Exact Solution =  (1.0, 9.88098e-14, -5.92415e-13, 1.0)
matrix_exp(A) = (1.0, 9.9e-14, -5.94e-13, 1.0)

The relative error for the small entries is very high. Pasted below is the C++ code:

```
    Matrix<double, Dynamic, Dynamic> m(2, 2), exp_m(2, 2), D(2, 2),
      D_inv(2, 2), A(2, 2), exp_A(2, 2);
    m << 1e-13, 0, 0, 1e-15;
    D << 1, 2, 3, 4;
    D_inv << -2, 1, 1.5, -0.5;
    exp_m << exp(1e-13), 0, 0, exp(1e-15);
    A << D * m * D_inv;
    
    std::cout << A << std::endl << std::endl;
    std::cout << D * exp_m * D_inv << std::endl << std::endl;
    std::cout << stan::math::matrix_exp(A) << std::endl << std::endl;
```

charlesm93 commented:
> How slow is it to test larger matrices?

The test for larger matrixes is slow for the rev case:
* 1629 ms for 10x10
* no response from terminal window after 10 minutes

For fwd:
* 1 ms for 10x10
* 72 ms for 100x100

> I also don't know if that's a general enough test.  You
> mean that the matrices you test are generated as S * B / S
> for some S and B?

Yes. I randomly generate S, making sure that it is invertible. Then I randomly generate the elements of the diagonal matrix B.

This is based on Moler & Van Loan, 2003. They use this strategy to construct a 2x2 matrix and show how a Taylor series fails to approximate the matrix exponential. Since Padé is also a series method, I expect it to be a robust test. Padé, unlike decomposition methods, does not take advantage of the fact A = S * B / S, so the test should be general enough.

bob-carpenter commented:
Even thought the inputs are are all on the order of 1e-14, the
outputs are order 1 and order 1e-14, which is probably where
the issue's arising.  So maybe relative error should be defined
relative to the entire set of matrix entries or maybe the diagonal
entries, in which case this is still low relative error.

> On Sep 28, 2016, at 2:44 PM, Charles Margossian <notifications@github.com> wrote:
> 
> ...
> A = (-1.97e-13, 9.9e-14, 5.94e-13, 2.98e-13)
> 
> Exact Solution = (1.0, 9.88098e-14, -5.92415e-13, 1.0)
> matrix_exp(A) = (1.0, 9.9e-14, -5.94e-13, 1.0)

bob-carpenter commented:
> On Sep 28, 2016, at 3:01 PM, Charles Margossian <notifications@github.com> wrote:
> 
...
> The test for larger matrixes is slow for the rev case:
> 	• 1629 ms for 10x10
> For fwd:
> 	• 1 ms for 10x10
> 	• 72 ms for 100x100
...

I assume that 72ms for fwd is not testing all the derivatives the
way the rev case is.  But yes, 1.6s is slow enough that we don't
want to go a lot higher than that in our usual test runs.  You could
stretch it out to 30s, but try not to exceed that for a test for
a single function.

- Bob

charlesm93 commented:
Here's how the relative error is now evaluated:
* The maximum absolute coefficient of the matrices we are comparing times 1e-10
* If we know the matrices have entries with roughly the same order of magnitude, use expect_matrix_eq (i.e. EXPECT_FLOAT_EQ).

The same principle applies to derivatives.

charlesm93 commented:
> I assume that 72ms for fwd is not testing all the derivatives the
> way the rev case is.

The input matrix A is a function of a set of parameters. In the fwd case, these parameters have d_ = 1, and essentially depend on an independent variable that is not explicitly declared. I'm testing for the derivative of the matrix_exponential with respect to that variable. 

In the rev case, every parameter is an independent variable. In addition -- and maybe I got confused on how the grad() function works -- the matrices get constructed for every entry of the output matrix (see the code for the 2x2 case below). My understanding is that the expression graph needs to be reinitialized every time we call grad(). Confirmation or clarification welcomed. 

But that explains why the rev case is computationally more intensive than the fwd case.

```
    for (size_t k = 0; k < 2; k++) {
    	for (size_t l = 0; l < 2; l++) {

    		matrix_v m1(2,2), m2(2,2), m1_exp;

    		AVAR a = -1.0, b = -17.0;

    		m1 << -2*a + 3*b, 1.5*a - 1.5*b, -4*a + 4*b, 3*a - 2*b;
   			m2 << -.735759, .551819, -1.471518, 1.103638;
    		m1_exp = stan::math::matrix_exp_pade(m1);
    		expect_matrix_eq(m2, m1_exp);

			matrix_v dm1_exp_da(2,2), dm1_exp_db(2,2);
    		dm1_exp_da << -2*exp(a), 1.5*exp(a), -4*exp(a), 3*exp(a);
    		dm1_exp_db << 3*exp(b), -1.5*exp(b), 4*exp(b), -2*exp(b);

    		AVEC x = createAVEC(a, b);
    		VEC g;
    		m1_exp(k, l).grad(x, g);
    		EXPECT_FLOAT_EQ(dm1_exp_da(k, l).val(), g[0]);
    		EXPECT_FLOAT_EQ(dm1_exp_db(k, l).val(), g[1]);
    	}
    }
```

bob-carpenter commented:
> On Sep 30, 2016, at 3:44 PM, Charles Margossian <notifications@github.com> wrote:
> 
> I assume that 72ms for fwd is not testing all the derivatives the
> way the rev case is.
> 
> The input matrix A is a function of a set of parameters. In the fwd case, these parameters have d_ = 1, and essentially depend on an independent variable that is not explicitly declared. I'm testing for the derivative of the matrix_exponential with respect to that variable.
> 
> In the rev case, every parameter is an independent variable. In addition -- and maybe I got confused on how the grad() function works -- the matrices get constructed for every entry of the output matrix (see the code for the 2x2 case below). My understanding is that the expression graph needs to be reinitialized every time we call grad(). Confirmation or clarification welcomed.

var.grad() works on a precomputed expression graph, so it's not
rebuilt.  But you do need to walk over the stack and reset
the adjoints if you're going to call it multiple times.
You can see the code for Jacobians in the math library to
see how this works:  

stan/math/rev/mat/functor/jacobian.hpp

I'd suggest studying the arXiv paper to see how our autodiff
works and then asking us questions based on that (it'll help
sharpen the paper, too).

- Bob

charlesm93 commented:
@bob-carpenter : Thanks for pointing me to jacobian.hpp and the autodiff paper. This allowed me to optimize the code, and test a 25x25 matrix in the same runtime.

In the previous version, instead of only resetting the adjoints, I redeclared all var (and matrix_v) variables, which was of course less efficient because this meant computing the values again.

I think the paper explains autodiff very clearly. I couldn't use the Jacobian function directly, since matrix_exp doesn't map to R^n (rather to R^n x R^n), but I followed its architecture closely.

I haven't said it in a while, but thank you for all the guidance.

charlesm93 commented on stan/math/prim/mat/fun/matrix_exp_pade.hpp:
```diff
@@ -0,0 +1,250 @@
+
+#include <stan/math/prim/mat/fun/Eigen.hpp>
+
+namespace stan {
```
It is from the beta version of Eigen 3.3 in unsupported (so not the currently implemented Eigen version in Stan), and has been modified to be auto-differentialble. Their functions were overloaded for float, double, and long double, but not for var and fvar. I created a template version.

bob-carpenter commented:
When the formatting's done, let me know and I can take a final code review
pass over everything.  Great that this is going in and sorry it's been
such a struggle---you took on some complicated examples to kick off!

- Bob

> On Oct 13, 2016, at 9:39 AM, Charles Margossian <notifications@github.com> wrote:
> 
> @charlesm93 commented on this pull request.
> 
> In stan/math/prim/mat/fun/matrix_exp_pade.hpp:
> 
> > +// Copyright (C) 2009, 2010, 2013 Jitse Niesen <jitse@maths.leeds.ac.uk>
> +// Copyright (C) 2011, 2013 Chen-Pang He <jdh8@ms63.hinet.net>
> +//
> +// This Source Code Form is subject to the terms of the Mozilla
> +// Public License v. 2.0. If a copy of the MPL was not distributed
> +// with this file, You can obtain one at http://mozilla.org/MPL/2.0/.
> +//
> +// This file was edited for to the Stan math library to create
> +// the matrix exponential function (matrix_exp), 2016.
> +
> +#ifndef STAN_MATH_PRIM_MAT_FUN_MATRIX_EXP_PADE_HPP
> +#define STAN_MATH_PRIM_MAT_FUN_MATRIX_EXP_PADE_HPP
> +
> +#include <stan/math/prim/mat/fun/Eigen.hpp>
> +
> +namespace stan {
> 
> It is from the beta version of Eigen 3.3 in unsupported (so not the currently implemented Eigen version in Stan), and has been modified to be auto-differentialble. Their functions were overloaded for float, double, and long double, but not for var and fvar. I created a template version.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

charlesm93 commented on stan/math/prim/mat/fun/matrix_exp_pade.hpp:
```diff
@@ -0,0 +1,250 @@
+namespace stan {
+    namespace math {
+
+        using Eigen::NumTraits;
```
Isn't Eigen::NumTraits a structure, as defined in eigen_3.2.9/Eigen/src/Core/NumTraits.h?

bob-carpenter commented:
Yes, Eigen::NumTraits is a class.   You can look these things
up in the Eigen doc.  It's pretty well indexed.

charlesm93 commented:
In that case, it should be ok to have `using Eigen::NumTraits` in the code, since it isn't a namespace level using statement (unless I am missing something, of course).

bob-carpenter commented:
That is a using statement declared at namespace scope.  I answered the question without looking the the use the first time.  We don't want using statements at the namespace level, no matter what they are.

charlesm93 commented:
Ok - now I get what Daniel meant.

bob-carpenter commented:
We'll be doing a release soon (probably in around 2 or 3 weeks time), so if you want this in 2.13.0, it should get wrapped up sooner rather than later with the accompanying stan-dev/stan issue to expose the function to the language.  I'm happy to help with doc.

No rush on our end if you'd rather wait.

charlesm93 commented:
The branch to expose matrix_exp() to Stan is Feature/issue-2043-matrix-exponential. I need to make sure all the tests are completed. The doc is for the manual?

charlesm93 commented:
Regarding documentation:

In `function.tex`, I am not sure where to put matrix_exp. The two options I see are \it{special matrix functions} (which seems ambiguous) and \it{linear algebra functions}. 

I'm also debating adding something to `programming.tex`, under Solving Differential Equations. Using Matrix Exp is much more efficient than numerical solvers for linear ODE's.

syclik commented on stan/math/rev/scal/fun/frexp.hpp:
```diff
@@ -0,0 +1,50 @@
+
+    /**
+	 * Variable implementation of frexp. The derivative of frexp
+	 * is 0, because frexp returns a discontinuous object.
```
@charlesm93, is the derivative 0 or NaN? In the `chain()` method, it's using NaN.

syclik commented:
@charlesm93, a few more comments:

1. This is cosmetic, but it matters: no tabs in code. The only place tabs are appropriate is in the makefiles.
2. This is substantial: please move the imported Eigen file into its own header file. Preferably called `MatrixExponential.h` like where it originally came from. I think it's ok to put all the functions in Eigen there. I'd actually prefer to leave it in the Eigen namespace. The `matrix_exp_pade.hpp` file should only include the last function: `matrix_exp_pade`.
3. What's with the C macro, `LDBL_MANT_DIG`?
4. Please remove the `//  end namespace math` comments.

syclik commented:
I'm also going to suggest we don't reimplement `frexp` for all the instantiations. I don't see how this function is going to be practically useful anywhere. If you're modifying the code, you can modify the call to the `frexp()` function to use doubles only. I'll add an inline comment.

syclik commented on stan/math/fwd/scal.hpp:
```diff
@@ -39,6 +39,7 @@
 #include <stan/math/fwd/scal/fun/fmax.hpp>
 #include <stan/math/fwd/scal/fun/fmin.hpp>
 #include <stan/math/fwd/scal/fun/fmod.hpp>
+#include <stan/math/fwd/scal/fun/frexp.hpp>
```
Remove this line.

syclik commented on stan/math/fwd/scal/fun/frexp.hpp:
```diff
@@ -0,0 +1,29 @@
+#ifndef STAN_MATH_FWD_SCAL_FUN_FREXP_HPP
```
Remove this file.

syclik commented on stan/math/prim/mat/fun/matrix_exp_pade.hpp:
```diff
@@ -0,0 +1,249 @@
+                    matrix_exp_pade9(arg, U, V);
+                } else {
+                    const double maxnorm = 5.371920351148152;
+                    frexp(l1norm / maxnorm, &squarings);
```
change this to:
```
frexp(value_of_rec(l1norm) / value_of_rec(maxnorm));
```

syclik replied:
Once you do that, you won't need the `frexp` specializations.

syclik commented on stan/math/rev/scal.hpp:
```diff
@@ -44,6 +44,7 @@
 #include <stan/math/rev/scal/fun/fmax.hpp>
 #include <stan/math/rev/scal/fun/fmin.hpp>
 #include <stan/math/rev/scal/fun/fmod.hpp>
+#include <stan/math/rev/scal/fun/frexp.hpp>
```
Remove this line.

syclik commented on stan/math/rev/scal/fun/frexp.hpp:
```diff
@@ -0,0 +1,50 @@
+#ifndef STAN_MATH_REV_SCAL_FUN_FREXP_HPP
```
remove this file.

bob-carpenter commented:
I was the one who suggested moving the Eigen function to an .hpp file and keeping it inline with our file naming conventions, but it won't bug me too much if it's the only .h file other than the signatures file.

charlesm93 commented:
@bob-carpenter : `MatrixExponential.h` is odd. On the one hand, it makes sense to include it in `matrix_exp_pade.hpp`, because `MatrixExponential.h` contains the code required to compute a pade approximation. On the other hand, we decided to minimally edit Eigen's original file. Keeping the file's original name and type is consistent with that. As a .h file, it is also not subject to cpplint's check.

I agree that one way or the other works fine. When we add Eigen 3 some of the functions in the structures of Unsupported/Eigen will become public, meaning I won't need to redefine these structures in `MatrixFunction.h` in order to construct a template version of `computeUV()`. I might move `computeUV()` to matrix_exp_pade.cpp then, and get rid of `MatrixFunction.h`, but that's hypothetical.

charlesm93 commented:
@syclik : 
- sorry about the tabs, I thought I had gotten rid of all them
- The C macro, as far as I understand, is for floating points with higher than quadruple precision. It triggers matrix functions of Eigen/unsupported (matrix_exp among others) to call so-called StdStemFunctions. Setting LDBL_MANT_DIG > 2, I realized the statements in the macro of statement do not compile. They rely on features from Eigen 3.3-beta. So I need to adjust them. But before I do, I wanted to ask: is it ever the case that LDBL_MANT_DIG > 112 ? On my local machine LDBL_MANT_DIG = 64. This conditional only applies to double and long doubles, but I guess the precision would propagate to rev and fwd variables.

bob-carpenter commented:
I'm also OK either way, so let's go with what Daniel wants, which
is to keep the original name.  I think if it's a .h it won't
get cpplinted anyway, so maybe we don't need to modify it at all?

> On Oct 23, 2016, at 4:30 PM, Charles Margossian <notifications@github.com> wrote:
> 
> @bob-carpenter : MatrixExponential.h is odd. On the one hand, it makes sense to include it in matrix_exp_pade.hpp, because MatrixExponential.h contains the code required to compute a pade approximation. On the other hand, we decided to minimally edit Eigen's original file. Keeping the file's original name and type is consistent with that. As a .h file, it is also not subject to cpplint's check.
> 
> I agree that one way or the other works fine. When we add Eigen 3 some of the functions in the structures of Unsupported/Eigen will become public, meaning I won't need to redefine these structures in MatrixFunction.h in order to construct a template version of computeUV(). I might move computeUV() to matrix_exp_pade.cpp then, and get rid of MatrixFunction.h, but that's hypothetical.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

bob-carpenter commented:
Doc will be for the manual on the stan-dev/stan repo, not here.  This just needs the API doc for doxygen.

syclik commented:
Sorry, I realize that I wasn't clear. I don't care that the file is renamed. What I want is the Eigen file to be pulled into the code base with minimal changes and have a different file for our code. I don't want one file with both things in it. I think the plan is to remove the file from our code base once we have Eigen 3.3. I would like that step to be easy and less error prone. 



> On Oct 25, 2016, at 2:11 PM, Bob Carpenter <notifications@github.com> wrote:
> 
> I'm also OK either way, so let's go with what Daniel wants, which
> is to keep the original name. I think if it's a .h it won't
> get cpplinted anyway, so maybe we don't need to modify it at all?
> 
> > On Oct 23, 2016, at 4:30 PM, Charles Margossian <notifications@github.com> wrote:
> > 
> > @bob-carpenter : MatrixExponential.h is odd. On the one hand, it makes sense to include it in matrix_exp_pade.hpp, because MatrixExponential.h contains the code required to compute a pade approximation. On the other hand, we decided to minimally edit Eigen's original file. Keeping the file's original name and type is consistent with that. As a .h file, it is also not subject to cpplint's check.
> > 
> > I agree that one way or the other works fine. When we add Eigen 3 some of the functions in the structures of Unsupported/Eigen will become public, meaning I won't need to redefine these structures in MatrixFunction.h in order to construct a template version of computeUV(). I might move computeUV() to matrix_exp_pade.cpp then, and get rid of MatrixFunction.h, but that's hypothetical.
> > 
> > —
> > You are receiving this because you were mentioned.
> > Reply to this email directly, view it on GitHub, or mute the thread.
> > 
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

bob-carpenter commented:
Either way is fine.  This will be on the stan-dev/stan issue anyway.  Sorry for the confusion.

bgoodri commented:
I would put it near the eigenvalues section but presumably in its own
section. And reference it from the differential equations section.

On Fri, Oct 21, 2016 at 10:01 AM, Charles Margossian <
notifications@github.com> wrote:

> Regarding documentation:
>
> In function.tex, I am not sure where to put matrix_exp. The two options I
> see are \it{special matrix functions} (which seems ambiguous) and
> \it{linear algebra functions}.
>
> I'm also debating adding something to programming.tex, under Solving
> Differential Equations. Using Matrix Exp is much more efficient than
> numerical solvers for linear ODE's.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/stan-dev/math/pull/390#issuecomment-255384935>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ADOrqursx9W2LNb82WlpRiXiPo_REp7gks5q2MWtgaJpZM4J5VLn>
> .
>

charlesm93 commented:
Moving the conversation on the docs to https://github.com/stan-dev/stan/pull/2092.

charlesm93 commented:
I sent an e-mail to the Eigen developers about the macro. I've looked at `result = arg.matrixFunction(stem_function_exp);` and I don't see what `stem_function_exp` does to address the higher precision case (or how it computes a matrix exponential for that matter).

The function is defined in `StemFunction.h`:

```
template <typename Scalar>
Scalar stem_function_exp(Scalar x, int)
{
  using std::exp;
  return exp(x);
}
```

bob-carpenter commented:
That code will use argument dependent lookup (ADL --- you
can find discussions online as to how it works) to find
an implementation for the Scalar type.  So if you
couple this with an implementation of exp, for
instance as in Stan's for autodiff, then this version
will be usable whereas a direct call to std::exp(x) wouldn't.

- Bob

> On Oct 31, 2016, at 10:14 AM, Charles Margossian <notifications@github.com> wrote:
> 
> I sent an e-mail to the Eigen developers about the macro. I've looked at result = arg.matrixFunction(stem_function_exp); and I don't see what stem_function_exp does to address the higher precision case (or how it computes a matrix exponential for that matter).
> 
> The function is defined in StemFunction.h:
> 
> template <typename Scalar>
> Scalar stem_function_exp(Scalar x, int)
> {
>   using std::exp;
>   return exp(x);
> }
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

charlesm93 commented:
When LDBL_MANT_DIG > 112, the matrix function uses a different method to compute the matrix exponential. In Gael Guennebaud from Eigen's words:

> In the general case it basically computes the eigenvalue/vector decomposition, apply the scalar-wise exp function (stem_function_exp) to eigenvalues, and then recompose the matrix. In practice, this is performed using a Schur decomposition instead of a full eigenvalue/vector decomposition.

When LDBL_MANT_DIG <= 112 on the other hand, the function uses a Pade approximation. 

For this pull request, I propose to only use Pade's algorithm, and write a cautionary note in the docs for users running on LDBL_MANT_DIG > 112. I will also submit issues on (1) implementing Schur decomposition and (2) adding the Schur decomposition algorithm to the matrix_exp function.

Honestly, I don't think getting the LDBL_MANT_DIG > 112 case is a priority. Under the current algorithm, all unit tests pass with LDBL_MANT_DIG = 128.

bob-carpenter commented:
If you know the Pade algorithm is going to go off the rails, you can throw an exception at run time based on input size or some other measurable piece of the algorithm.

Once you have derivatives implemented, then you may have to split those, too.

bgoodri commented:
I have the real Schur decomposition on a branch.

On Wed, Nov 2, 2016 at 3:17 PM, Bob Carpenter <notifications@github.com>
wrote:

> If you know the Pade algorithm is going to go off the rails, you can throw
> an exception at run time based on input size or some other measurable piece
> of the algorithm.
>
> Once you have derivatives implemented, then you may have to split those,
> too.
>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/stan-dev/math/pull/390#issuecomment-257970369>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ADOrqtck3Usbuy6aMNmSKLPVV0CacmMtks5q6OGygaJpZM4J5VLn>
> .
>

charlesm93 commented:
@bob-carpenter : I'm not 100% sure if there is a method to predict the algorithm will go off the rail. Size alone won't do it. There is a theoretical upper-bound for the relative error (Moler & Von Loan, section 3, method 3), which depends on ||A||, the degree of the Pade approximation _j_ and the squaring _q_ (the latter two are automatically picked by the matrix_exp function). 

Computing the upper bound is not hard per se, but it might be computationally expensive. One has:

rel_err <= eps * ||A|| * exp(eps * ||A||)

where eps = 8 * ( ||A|| / 2^j )^(2q) * ( (q!)^2 / ( (2q)! * (2q + 1)! )

I can imagine requiring the user to specify rel_err, kind of like we do for ODE integrators. This is overkill for pharmacometrics problem, but I can't speak for other fields. I'm putting this on the agenda for the next Stan meeting.

charlesm93 commented:
> If you know the Pade algorithm is going to go off the rails, you can throw an exception at run time based on input size or some other measurable piece of the algorithm.

Scaling and squaring takes care of errors due to large ||A|| or widely spread eigenvalues [1]. Squaring however introduces a round-off error. Numerical experiments show that the relative roundoff error can be estimated using the upper-bound: **u**_n_||A||_1, where **u** is the computer relative precision, _n_ the order of the matrix, and ||A||1 the 1-norm of A [2].

For a C++ double, **u** ~ e-16. Thus for a 100 x 100 matrix, where each entry is of the order e+2, the upper bound would be: eps ~ e-16 * e+2 * e+2 * e+2 = e-10, which is acceptable. Computing this *empirical* upper bound shouldn't be too expensive, because Eigen's code already computes ||A||_1 when determining the degree of the pade approximation. _n_ is trivial, and **u** is a constant.

I can add the exception. I am reluctant to do so because it seems overkill; on the other hand, the cost should not be too high.

Is there a standard relative error implemented in Stan? Is this something the user should specify (in my opinion, no)?

[1] Roger B. Sidje, Expokit: A Software Package for Computing Matrix Exponentials, 1998
[2] M. Arioli, B. Codenotti, & C. Fassino, The Pade Method for Computing the Matrix Exponential from Linear Algebra and Its Applications, 1996

bob-carpenter commented:
I kicked off the failing Travis job again.  If that passes, I think we're good to merge.

bob-carpenter commented on stan/math/prim/mat/fun/matrix_exp_2x2.hpp:
```diff
@@ -0,0 +1,45 @@
+        matrix_exp_2x2(
+          const Eigen::Matrix<T, Eigen::Dynamic, Eigen::Dynamic>& A) {
+            T a = A(0, 0), b = A(0, 1), c = A(1, 0), d = A(1, 1), delta;
+            delta = sqrt((a - d) * (a - d) + 4 * b * c);
```
same here.  use `square(a - d)`.

bob-carpenter commented on stan/math/prim/mat/fun/matrix_exp_2x2.hpp:
```diff
@@ -0,0 +1,45 @@
+            B(0, 0) = exp_half_a_plus_d * (delta_cosh + ad_sinh_half_delta);
+            B(0, 1) = b * Two_exp_sinh;
+            B(1, 0) = c * Two_exp_sinh;
+            B(1, 1) = exp_half_a_plus_d * (delta_cosh - ad_sinh_half_delta);
```
Save this repeated term `delta_cosh + ad_sinh_half_delta` and reuse.  Basic rule is to never repeat a term.  It just leads to bigger-than-necessary autodiff stacks.

bob-carpenter replied:
Ack!  I missed that.  Good thing I'm not just trying to change these myself.

bob-carpenter commented on stan/math/prim/scal/fun/ldexp.hpp:
```diff
@@ -0,0 +1,25 @@
+      */
+    template <typename T>
+    inline T ldexp(const T& a, int b) {
+      return a * std::pow(2.0, b);
```
use the `exp2` function (built into Stan)

bob-carpenter commented:
Sorry---had some more comments.

bob-carpenter commented on stan/math/prim/scal/fun/ldexp.hpp:
```diff
@@ -17,7 +17,7 @@ namespace stan {
     template <typename T>
     inline T ldexp(const T& a, int b) {
-      return a * std::pow(2.0, b);
+      return a * exp2(b);
```
This will need the include for exp2.  Run the header tests.

charlesm93 replied:
To run the header tests, use `make test-math-dependencies`, right? When I corrected for the error message it gave me, the next time I tried to ran it, 'nothing to be done...' Is this an expected behavior?

charlesm93 commented on stan/math/prim/mat/fun/matrix_exp_pade.hpp:
```diff
@@ -0,0 +1,39 @@
+          MatrixType numer = U + V;
+          MatrixType denom = -U + V;
+          result = denom.partialPivLu().solve(numer);
+          for (int i = 0; i < squarings; i++)
```
Is there a reason for preferring one to the other?

bob-carpenter commented:
@sycklik are you OK with this now?  It looks good to me and I can merge when the unit tests pass.

syclik commented on stan/math/prim/mat/fun/MatrixExponential.h:
```diff
@@ -0,0 +1,211 @@
+#define STAN_MATH_PRIM_MAT_FUN_MATRIXEXPONENTIAL_H
+
+#include <stan/math/prim/scal/fun/value_of_rec.hpp>
+#include <stan/math/rev/scal/fun/value_of_rec.hpp>
```
this is in `prim`, so please remove the includes for `rev` and `fwd`.

syclik commented on stan/math/prim/mat/fun/MatrixExponential.h:
```diff
@@ -0,0 +1,211 @@
+  template <typename MatrixType>
+  struct matrix_exp_computeUV<MatrixType>
+  {  
+  	template <typename T>  
```
remove tab. Was the tab introduced by you or was this in Eigen?

charlesm93 replied:
my fault. I added this template a while ago, and forgot to change the tab to two spaces.

charlesm93 commented:
I don't understand the error Jenkins throws in the default test: `use of undeclared identifier 'std'` in `matrixExponential.h`. The changes in the last commit couldn't cause this error, right?

syclik commented:
@charlesm93: no idea, but this isn't a fluke of an issue. You can verify by running:
```
make test-headers
```

Or just the test that failed
```
make stan/math/prim/mat/fun/matrix_exp_pade.hpp-test
```

charlesm93 commented:
@syclik Yes, this confirms that the last commit causes the error. Putting back the rev and fwd include statements prevents the error. Really odd. Time for some digging.

syclik commented:
Just looked at it. Can you try putting in the include for the prim/mat version of the value_of_rec function?

> On Nov 14, 2016, at 3:22 PM, Charles Margossian <notifications@github.com> wrote:
> 
> @syclik Yes, this confirms that the last commit causes the error. Putting back the rev and fwd include statements prevents the error. Really odd. Time for some digging.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

charlesm93 commented:
ok, I'm on it

charlesm93 commented:
A solution that works for the matrix_exp function is to include <Eigen/Dense> in prim/.../value_of_rec. The other option would be to include <stan/math/prim/mat/fun/Eigen.hpp>, which includes two more files from Eigen. Not needed, but consistent with what rev/.../value_of_rec does.

syclik commented on stan/math/prim/scal/fun/value_of_rec.hpp:
```diff
@@ -1,6 +1,8 @@
 #ifndef STAN_MATH_PRIM_SCAL_FUN_VALUE_OF_REC_HPP
 #define STAN_MATH_PRIM_SCAL_FUN_VALUE_OF_REC_HPP
 
+#include <Eigen/Dense>
```
@charlesm93: things inside `prim` cannot bring in any of Eigen's headers. This header include needs to be removed from this file.

syclik commented on stan/math/prim/mat/fun/MatrixExponential.h:
```diff
@@ -15,7 +15,7 @@
 
 #include <stan/math/prim/scal/fun/value_of_rec.hpp>
-#include <Eigen/Dense>
+// #include <Eigen/Dense>
```
Remove this commented line.

syclik commented on stan/math/prim/mat/fun/MatrixExponential.h:
```diff
@@ -15,7 +15,7 @@
 #define STAN_MATH_PRIM_MAT_FUN_MATRIXEXPONENTIAL_H
 
 #include <stan/math/prim/scal/fun/value_of_rec.hpp>
```
Change this to include:
`stan/math/prim/mat/fun/value_of_rec.hpp`

Note: you should be including the `mat` version, not the `scal` version.

charlesm93 replied:
Ah yes, that makes sense, and it explains why the header test didn't work.
